CitationStyleLanguagePlugin.php

```diff
@@ -0,0 +1,1048 @@
+    {
+        return $this->getSetting($contextId, 'groupAuthor') ?? [];
+    }
+
+    public function getChapterAuthorGroups(int $contextId = 0): array
+    {
+        return $this->getSetting($contextId, 'groupChapterAuthor') ?? [];
+    }
+
+    public function getSerialNumber(Publication $publication): array
+    {
+        $serialNumber = [];
+        $publicationFormats = $publication->getData('publicationFormats');
+        /** @var PublicationFormat $publicationFormat */
+        foreach ($publicationFormats as $publicationFormat) {
+            if ($publicationFormat->getIsApproved()) {
+                $identificationCodes = $publicationFormat->getIdentificationCodes();
+                $identificationCodes = $identificationCodes->toArray();
+                foreach ($identificationCodes as $identificationCode) {
+                    $serialNumber[] = htmlspecialchars($identificationCode->getValue());
+                }
+            }
+        }
+        return $serialNumber;
+    }
+
+    protected function addSeriesInformation(stdClass $citationData, Publication $publication): stdClass
+    {
+        $seriesId = $publication->getData('seriesId');
+        if ($seriesId) {
+            /** @var SeriesDAO $seriesDao */
+            $seriesDao = DAORegistry::getDAO('SeriesDAO');
+            if (null !== $seriesDao) {
+                $series = $seriesDao->getById($seriesId);
+                if (null !== $series) {
+                    $citationData->{'collection-title'} = htmlspecialchars(trim($series->getLocalizedFullTitle()));
+                    $citationData->volume = htmlspecialchars($publication->getData('seriesPosition'));
+                    $citationData->{'collection-editor'} = htmlspecialchars($series->getEditorsString());
+                    $onlineISSN = $series->getOnlineISSN();
+                    if (!empty($onlineISSN)) {
+                        $citationData->serialNumber[] = htmlspecialchars($onlineISSN);
+                    }
+                    $printISSN = $series->getPrintISSN();
+                    if (!empty($printISSN)) {
+                        $citationData->serialNumber[] = htmlspecialchars($printISSN);
+                    }
+                }
+            }
+        }
+        return $citationData;
+    }
+
+    protected function getChapter(PKPRequest $request, Publication $publication): ?Chapter
+    {
+        $args = $request->getRequestedArgs();
+
+        if (isset($args[1], $args[2]) && $args[1] === 'chapter') {
+            $key = 2;
+        } elseif (isset($args[1], $args[3], $args[4]) && $args[1] === 'version' && $args[3] === 'chapter') {
+            $key = 4;
+        } else {
+            $key = 0;
+        }
+
+        if ($key !== 0) {
+            $chapterId = (int) $args[$key];
+            if ($chapterId > 0) {
+                $chapterDao = DAORegistry::getDAO('ChapterDAO');
+                $chapters = $chapterDao->getBySourceChapterId($chapterId);
```

Comment:
I think this might be wrong -- the source chapter ID will be (if I remember right) the ID of the chapter from a previous version (publication), no? I think this should be `$chapterDao->getChapter($chapterId, $publication->getId()` (which does the publication ID check at the same time for free).

Reply:
The source chapter ID is the ID from the first version of a chapter. In the url for a chapter landing page we use always the source chapter ID. This is important for DOI registration of a chapter. So frontend useres should never see an other chapter ID as the source chapter ID. So we get from `$request->getRequestedArgs()` always the source chapter ID.

Maybe it would be better to change the function `getBySourceChapterId($chapterId)` to take a optional second argument publication ID. Then we could put this code in  ChapterDAO.

Reply:
Ah! I'd forgotten about this element of OMP. You're right, the source chapter ID is the correct thing to use. Yes, if you don't mind, please do propose a change to add the publication ID to `getBySourceChapterId` as an optional second parameter, thanks!

Reply:
Okay, it's done.